shotover-proxy/src/codec/cassandra.rs

```diff
@@ -113,14 +76,42 @@ impl Decoder for CassandraCodec {
                     }
                 }
                 Err(CheckEnvelopeSizeError::UnsupportedVersion(version)) => {
-                    return Ok(Some(vec![self.reject_protocol_version(version)]));
+                    return Err(reject_protocol_version(version));
+                }
+                err => {
+                    return Err(CodecReadError::Parser(anyhow!(
+                        "Failed to parse frame {:?}",
+                        err
+                    )))
                 }
-                err => return Err(anyhow!("Failed to parse frame {:?}", err)),
             }
         }
     }
 }
 
+/// If the client tried to use a protocol that we dont support then we need to reject it.
+/// The rejection process is sending back an error and then closing the connection.
+fn reject_protocol_version(version: u8) -> CodecReadError {
+    info!(
+        "Negotiating protocol version: rejecting version {} (configure the client to use a supported version by default to improve connection time)",
+        version
+    );
+
+    CodecReadError::RespondAndThenCloseConnection(vec![Message::from_frame(Frame::Cassandra(
+        CassandraFrame {
+            version: Version::V4,
+            stream_id: 0,
+            operation: CassandraOperation::Error(ErrorBody {
+                error_code: 0xA, // https://github.com/apache/cassandra/blob/adf2f4c83a2766ef8ebd20b35b49df50957bdf5e/doc/native_protocol_v4.spec#L1053
```

Comment:
Is this the right error code to throw for an unsupported version? The spec just says this if for a protocol error, an unsupported version may technically not be a protocol error itself?

What happens when we do a pcap on a bad version negotation with a direct connection to cassandra (or what does the cassandra code do)

Reply:
A ProtocolException is thrown for invalid versions: https://github.com/apache/cassandra/blob/cd576a0d140c94007fd0181044df453c0681252f/src/java/org/apache/cassandra/transport/ProtocolVersion.java#L113

A ProtocolException corresponds to PROTOCOL_ERROR https://github.com/apache/cassandra/blob/cd576a0d140c94007fd0181044df453c0681252f/src/java/org/apache/cassandra/transport/ProtocolException.java#L43

Which is the magic 0xA int: https://github.com/apache/cassandra/blob/cd576a0d140c94007fd0181044df453c0681252f/src/java/org/apache/cassandra/exceptions/ExceptionCode.java#L35